[GPII-3743]: Improve PDs handling in couchdb module

natarajaya opened the pull request:
This implements the following:
* Adds variable to enable / disable PVCs recovery code and disables it for dev environment.
* ~~Adds code to remove disks with `gcloud` after PVCs deletion to not leave stale resources behind.~~

natarajaya commented:
I believe this also solves GPII-3742.

amatas commented:
LGTM

mrtyler commented on gcp/live/dev/k8s/gpii/couchdb/terraform.tfvars:
```diff
@@ -31,3 +31,4 @@ pv_reclaim_policy    = "Delete"
 pv_storage_class     = ""
 pv_provisioner       = ""
 execute_destroy_pvcs = "true"
+execute_recover_pvcs = "false"
```
This change means that we will almost never test the `execute_recover_pvcs` workflow. Does that concern you?

natarajaya replied:
Good question. We recreate staging cluster from time to time, so recover code is tested at that time, I guess this is better than never.

stepanstipl commented on gcp/modules/couchdb/main.tf:
```diff
@@ -150,6 +152,11 @@ resource "null_resource" "couchdb_destroy_pvcs" {
         done
+        # Some times GCP disks persist even after PVC destruction, removing them with gcloud to not leave stale resources behind
+        # More info: https://issues.gpii.net/browse/GPII-3743
+        for DISK in $(gcloud compute disks list --filter description:couchdb --format json | jq -r -c .[].name); do
```
Quoting the `.[].name` bit would make me feel easier, as well as the output of the whole `$()`, otherwise this might be expanded by shell in case of some unexpected output.

natarajaya replied:
Nice catch!

stepanstipl commented on gcp/modules/couchdb/main.tf:
```diff
@@ -150,6 +152,11 @@ resource "null_resource" "couchdb_destroy_pvcs" {
+        # Some times GCP disks persist even after PVC destruction, removing them with gcloud to not leave stale resources behind
+        # More info: https://issues.gpii.net/browse/GPII-3743
+        for DISK in $(gcloud compute disks list --filter description:couchdb --format json | jq -r -c .[].name); do
+          timeout -t 600 gcloud beta compute disks delete $DISK --quiet --region ${var.infra_region} 2> /dev/null || true
```
Why is the `beta` needed here?

natarajaya replied:
It allows to delete disks by providing region instead of zone. There is no easy way to extract zone from `disks list` json output.

natarajaya commented:
> Any idea why the underlying disks are not getting deleted?

@stepanstipl Unfortunately no – destroying PVCs supposed to release the disks, but in some small number of cases it is not happening. I was unable to find anything related to that event in cluster logs. And that scenario is also very hard to reproduce.

> Have a scenario when trying to delete disk that is still mounted, because it was not released by Kube

Deleting corresponding PVC releases the disk. 

> I'm slightly worried that we're introducing "workaround" to an issue which's cause we don't understand very well

Yeah, it is a good reason [to be slightly worried](https://github.com/gpii-ops/gpii-infra/blame/master/gcp/modules/couchdb/main.tf#L151). I have nothing better to offer for now.

> I disagree with this - two separate issues. This does not fix recovery process in any way

This PR disables recovery process for dev environments. I thought we agreed that for staging and prod we will follow the process of migrating disks to proper zones manually. But I am fine treating GPII-3742 separately.

natarajaya commented:
@stepanstipl Ok, I addressed some of your concerns: new code does not use `gcloud beta` anymore, it also repeats disks deletion in a loop until success is achieved.

stepanstipl commented on gcp/modules/couchdb/main.tf:
```diff
@@ -142,15 +144,41 @@ resource "null_resource" "couchdb_enable_pv_backups" {
+      while [ "$DISKS_DELETED" != "true" ]; do
+        DISKS_DELETED="true"
+        echo "[Try $RETRY_COUNT of $RETRIES] Deleting CouchDB disks..."
+        for DISK in $(gcloud compute disks list --filter description:couchdb --format json | jq -c '.[]'); do
```
Missing `""` on `$()`

stepanstipl commented on gcp/modules/couchdb/main.tf:
```diff
@@ -142,15 +144,41 @@ resource "null_resource" "couchdb_enable_pv_backups" {
+          sleep 10
+        fi
+        RETRY_COUNT=$(($RETRY_COUNT+1))
+      done
```
Can we get rid of the `exit` clause and move all the loop control logic to the while clause? I think it less prone to errors and makes it more obvious what controls the loop. 

Smth. like:
```bash
while [ "$RETRY_COUNT" -le "$RETRIES" -a "DISKS_NOT_DELETED" != '0' ]; then
  DISKS_NOT_DELETED=0
  for DISK ...; do
    ..
    if [ "$?" != '0' ]; then
      DISKS_NOT_DELETED=$((DISKS_NOT_DELETED+1))
    fi
  done

  # Sleep only if we're not done nor last run    
  if [ "$RETRY_COUNT" -lt "$RETRIES" -a "DISKS_DELETED" != '0' ]; then
    echo "Failed to delete #{DISKS_NOT_DELETED} disks"
    sleep 10
  fi

  RETRY_COUNT=$((RETRY_COUNT+1))
done  
```

natarajaya replied:
But we would still need to exit with code 1 to fail the build if disks are not deleted and we run out of attempts. Added missing condition to retry check.

natarajaya replied:
I'll follow your advice and remove disk deletion code instead.

natarajaya commented on gcp/modules/couchdb/main.tf:
```diff
@@ -142,15 +144,41 @@ resource "null_resource" "couchdb_enable_pv_backups" {
+      while [ "$DISKS_DELETED" != "true" ]; do
+        DISKS_DELETED="true"
+        echo "[Try $RETRY_COUNT of $RETRIES] Deleting CouchDB disks..."
+        for DISK in $(gcloud compute disks list --filter description:couchdb --format json | jq -c '.[]'); do
```
It will not work with quotes:
```bash
# for i in "1 2 3"; do echo $i; echo "check"; done
1 2 3
check

# for i in 1 2 3; do echo $i; echo "check"; done
1
check
2
check
3
check
```

natarajaya replied:
I don't think I understand your example. I tried number of scenarios with unexpected input from `gcloud` – when `jq` can not extract expected fields from json, it leaves `stdout` empty, so execution flow never reaches the loop body.

natarajaya commented on gcp/modules/couchdb/main.tf:
```diff
@@ -142,15 +144,41 @@ resource "null_resource" "couchdb_enable_pv_backups" {
+        echo "[Try $RETRY_COUNT of $RETRIES] Deleting CouchDB disks..."
+        for DISK in $(gcloud compute disks list --filter description:couchdb --format json | jq -c '.[]'); do
+          echo "Trying to delete $(echo $DISK | jq -r .name)..."
+          timeout -t 60 gcloud compute disks delete $(echo $DISK | jq -r .selfLink) --quiet 2> /dev/null
```
Same here.

natarajaya commented:
> So that is exactly my point - if the disk is never released and never detached from the VM, this fix won't work

> But overall I would still like to see some evidence that this solves the issue.

I run `for i in {1..25}; do rake redeploy_module['k8s/gpii/couchdb']; done` in my dev cluster and it worked every single time, disks are getting destroyed on 2-3rd attempt in average.

Also, main change here is not PD hard-removal code (it is mostly needed to not leave stale disks behind), but the change that disables PD recovery code for dev environments.

stepanstipl commented:
^ I've seen one more fail on recovery code, trying to recover disk that did not exist since then

natarajaya commented:
> I've seen one more fail on recovery run, trying to recover disk that did not exist since then

It looks like you set both `execute_destroy_pvcs` and `execute_recover_pvcs` into true? I don't think we ever want to do that, and of course there may be situation when recovery code is trying to access disks that are in the process of deletion by Kubernetes.

> So I think this is actually dangerous because we're deleting a disk that is managed by Kubernetes, before it has a chance to delete it on its own.

I did not find any side effects of forcefully deleting disks that got their corresponding PVCs and PVs removed and already scheduled for deletion.

> Maybe we can get just that change in, without introducing the other bits for now?

I am okay with this, the downside of course would be leaving stale disks behind.

natarajaya commented on gcp/modules/couchdb/main.tf:
```diff
@@ -2,6 +2,7 @@ terraform {
   backend "gcs" {}
 }
 
+variable "infra_region" {}
```
It was needed in hard-deletion code. Removed.

natarajaya commented:
Updated PR description.

stepanstipl commented:
lgtm 👍

amatas commented:
LGTM!
